### PR TITLE
Adding support for SafeSideEffect

### PR DESCRIPTION
**Pull request by hiteshsharma**

### CODE REVIEW

**Comment by hiteshsharma** [comment]
@mchen07 please review and accept whenever possible

File: subprojects/parseq/src/main/java/com/linkedin/parseq/Task.java [hunk 1]
```diff
@@ -359,6 +359,40 @@
...
+   * Differs in exception handling behavior from {@code withSideEffect(desc, func)}.
+   * If an exception is thrown in {@link Function1}(passed in arguments) then {@code safeSideEffect } will convert it
+   * into a {@code Task.failure } and will ensure that the exception doesn't spread it's impact outside the plan
+   * compared to {@code withSideEffect} in which an exception is thrown in {@link Function1}, the entire plan fails.
+   *
+   * @param desc description of the side Effect. This will show up in trace
+   * @param func function to be applied to get side effect task
+   * @return a new Task that will run the side effect Task
+   */
+  default Task<T> safeSideEffect(final String desc, final Function1<? super T, Task<?>> func) {
```
Comment (mchen07) [inline subprojects/parseq/src/main/java/com/linkedin/parseq/Task.java]: There are two versions of `withSideEffect` method, one taking `Function1`, the other taking `Callable`. We may also need to guard for `Callable` version.

File: subprojects/parseq/src/test/java/com/linkedin/parseq/AbstractTaskTest.java [hunk 2]
```diff
@@ -297,6 +298,23 @@ public void testWithSideEffectFailure(int expectedNumberOfTasks) throws Exceptio
     assertEquals(countTasks(failure.getTrace()), expectedNumberOfTasks);
   }
 
+  @Test
+  public void testSafeSideEffectFailureInTaskProviderFunction() {
```
Comment (mchen07) [inline subprojects/parseq/src/test/java/com/linkedin/parseq/AbstractTaskTest.java]: Can we also add a test to verify the behavior of not using `safeSideEffect`?

**Comment by hiteshsharma** [comment]
@mchen07 updated as per comments

**Comment by hiteshsharma** [comment]
@mchen07 please merge if all looks good

**Pull request merged**